### DISCUSSION TITLE
Is it possible to send events to an invoked state machine without going through parent?

### DISCUSSION THREAD

**Discussion by martynchamberlin** [statelyai/xstate#3826] [Q&A]
Hey,

Been searching high and low and can't find anything on this, so thought I'd ask. 😅 Let's say I have a simple parent-child state machine scenario as described at https://xstate.js.org/docs/guides/communication.html#invoking-machines. Here's a quick copy/paste of those machines, for convenience. 

```ts
import { createMachine, interpret, send, sendParent } from 'xstate';

// Invoked child machine
const minuteMachine = createMachine({
  id: 'timer',
  initial: 'active',
  states: {
    active: {
      after: {
        60000: { target: 'finished' }
      }
    },
    finished: { type: 'final' }
  }
});

const parentMachine = createMachine({
  id: 'parent',
  initial: 'pending',
  states: {
    pending: {
      invoke: {
        src: minuteMachine,
        // The onDone transition will be taken when the
        // minuteMachine has reached its top-level final state.
        onDone: 'timesUp'
      }
    },
    timesUp: {
      type: 'final'
    }
  }
});
```

Let's say I want to send an event to the `minuteMachine`, but I don't want to have to go through the `parentMachine`. (I know `minuteMachine` doesn't have any `on` events defined yet, but work with me.) Is there a way I can do that? If I do this: 

```
const service = interpret(minuteMachine)
```

...it's a different instance from the one that's invoked by the parent machine. Sending events to it has no bearing on the instance invoked by the parent machine. 

Just wondering if there's any way I can have a "singleton" pattern or similar on this child machine, so it's reference-able somehow outside the context of its parentMachine. Would that be considered an anti-pattern?

**Comment by ChrisShank** [comment]
You can grab a reference to an invoked service via `state.children`.

```ts
const parentService = interpret(parentMachine);
const minuteService = parentService.state.children.get(“minute”)
```

make sure to add an id when invoking minuteMachine

```ts
invoke: {
  id: “minute”
  src: minuteMachine,
  onDone: 'timesUp'
}
```

**Comment by davidkpiano** [comment]
You can also make a makeshift "receptionist pattern" in v4: https://stackblitz.com/edit/js-vcmeg3?file=index.js

```js
import { createMachine, interpret } from 'xstate';
import { registry } from 'xstate/lib/registry';

// Userland code; not that scary!
const namedActors = {};
const simpleReceptionist = {
  get(name) {
    return namedActors[name];
  },
  set(name, actorRef) {
    namedActors[name] = actorRef;
  },
};

function registerAs(name) {
  return (ctx, ev, { state }) => {
    const actorRef = registry.get(state._sessionid);
    simpleReceptionist.set(name, actorRef);
  };
}

const m = createMachine({
  invoke: [
    {
      src: createMachine({
        entry: registerAs('notifier'),
        on: {
          message: {
            actions: (ctx, ev) => {
              console.log('notifier received', ev);
            },
          },
        },
      }),
    },
    {
      src: (ctx, ev) => (sendBack) => {
        setInterval(() => {
          simpleReceptionist.get('notifier')?.send({
            type: 'message',
            data: Math.random(),
          });
        }, 1000);
      },
    },
  ],
});

interpret(m).start();
```

**Comment by martynchamberlin** [comment]
@davidkpiano Interesting idea. I would love this, except that `predictableActionArguments: true,` causes the `meta.state` property to no longer be present. Is there a way to do this in a way that will be compatible with v5? :neckbeard: If I add `predictableActionArguments: true` to your example, I get this error: 

> Error: Cannot read properties of undefined (reading '_sessionid')
